Deploy 2 instances of Prometheus

philandstuff opened the pull request:
Prometheus's fault-tolerance is based on it being a really dumb simple
thing to operate.  It doesn't do replication, backups, leadership
elections, distributed locking, or anything like that.  It's just a
single instance with a local disk.

As a result, if we lose the AZ that our single Prometheus is running
in, we'll lose Prometheus.  To guard against that, we should maybe
have 2 of them?

In a multiple Prometheus setup, each Prometheus is completely
independent.  Each Prometheus will scrape each target and maintain a
separate tsdb of time series data.  As a result, it's probably best to
always target the same Prometheus instance where possible so that you
don't get weird jitter in your results where different Prometheis
scraped out of phase with each other.

Therefore, this also sets the Prometheus service session affinity to
`ClientIP` which enforces sticky sessions. See
https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies
for more on this.

I considered making this configurable on a per-cluster basis but I
thought that was probably more effort than it was worth.  2 is a
sensible default.

philandstuff committed:
Deploy 2 instances of Prometheus

Prometheus's fault-tolerance is based on it being a really dumb simple
thing to operate.  It doesn't do replication, backups, leadership
elections, distributed locking, or anything like that.  It's just a
single instance with a local disk.

As a result, if we lose the AZ that our single Prometheus is running
in, we'll lose Prometheus.  To guard against that, we should maybe
have 2 of them?

In a multiple Prometheus setup, each Prometheus is completely
independent.  Each Prometheus will scrape each target and maintain a
separate tsdb of time series data.  As a result, it's probably best to
always target the same Prometheus instance where possible so that you
don't get weird jitter in your results where different Prometheis
scraped out of phase with each other.

Therefore, this also sets the Prometheus service session affinity to
`ClientIP` which enforces sticky sessions. See
https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies
for more on this.

I considered making this configurable on a per-cluster basis but I
thought that was probably more effort than it was worth.  2 is a
sensible default.

Krenair commented:
Does this guarantee Prometheus will get deployed to two different AZs?

chrisfarms reviewed (dismissed):
sounds like a plan.

`ClientIP` session afinity however sets off my istio-spidey-sense ... `gsp-system` is not currently in the mesh, so this is probably not an issue for now, but this sounds like something that may not play nice with istio's traffic management when we come to bring gsp-system into mesh land

philandstuff commented:
> Does this guarantee Prometheus will get deployed to two different AZs?

I haven't checked but I assume there's some sort of pod anti-affinity rules that require distinct values of `failure-domain.beta.kubernetes.io/zone`

philandstuff commented:
@Krenair It looks like the statefulset should automatically spread prometheus servers across zones: https://kubernetes.io/docs/setup/best-practices/multiple-zones/

However I have just discovered the default behaviour for prometheus is to use an `emptyDir` volume for data rather than a persistentvolume. So I'm going to fix that too.

philandstuff committed:
fix comment to be up to date

philandstuff committed:
Move prometheus to persistent storage

This provisions 50Gi worth of data for each prometheus.

Currently, the verify prometheus uses about 6Gi of data.  50Gi was an
arbitrary number I chose based on:

 - lots of headroom, so that we avoid resizing for as long as possible
   - we have docs on how to resize and it's not pretty:
     https://github.com/alphagov/gsp/blob/master/docs/gds-supported-platform/resizing-persistent-volumes.md
 - reasonable costs
   - EBS gp2 in London costs $0.116 per GB-month.  So this whole PR
     would cost (3 clusters, 2 prometheis per cluster, 50 GB per
     prometheus) 3*2*50*0.116 = $34.80 per month

philandstuff commented:
I've updated this PR to migrate prometheus to persistent volume storage.  Commit message c1506cd inlined here for visibility:

This provisions 50Gi worth of data for each prometheus.

Currently, the verify prometheus uses about 6Gi of data.  50Gi was an
arbitrary number I chose based on:

 - lots of headroom, so that we avoid resizing for as long as possible
   - we have docs on how to resize and it's not pretty:
     https://github.com/alphagov/gsp/blob/master/docs/gds-supported-platform/resizing-persistent-volumes.md
 - reasonable costs
   - EBS gp2 in London costs $0.116 per GB-month.  So this whole PR
     would cost (3 clusters, 2 prometheis per cluster, 50 GB per
     prometheus) 3*2*50*0.116 = $34.80 per month
